{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Querying population frequency using excel sheet as the input\n",
    "> The excel sheet should have column name of :\n",
    ">1. \"Coordinate\" for nucleotide position\n",
    ">2. \"Chr\" for Chromosomal number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os, sqlite3, time, datetime, subprocess\n",
    "import pandas as pd\n",
    "import multiprocessing\n",
    "import threading\n",
    "from threading import BoundedSemaphore\n",
    "from utils import logDecorator as lD\n",
    "\n",
    "from utils.createDB import *\n",
    "from utils.query import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Creating sql table for the input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/leehongkai/anaconda/lib/python3.6/site-packages/pandas/core/generic.py:1362: UserWarning: The spaces in these column names will not be changed. In pandas versions < 0.14, spaces were converted to underscores.\n",
      "  chunksize=chunksize, dtype=dtype)\n"
     ]
    }
   ],
   "source": [
    "xls_file = pd.ExcelFile('data/query.xlsx')\n",
    "table = xls_file.parse('Sheet1')\n",
    "# chrom = [x for x in table['Chr'].unique()[:-1]]\n",
    "conn = sqlite3.connect('dbase_Sqlite')\n",
    "try:\n",
    "    cleardb('dbase_Sqlite', 'dataFrame')\n",
    "except:\n",
    "    pass\n",
    "table.to_sql('dataFrame', conn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Querying from Chinese, Malay and Indian populations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 Single thread/cpu "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2017/11/30 09:55:51 - Data Retrieval for all races completed: Took 625.309369802475 seconds to complete.\n"
     ]
    }
   ],
   "source": [
    "startTime = time.time()\n",
    "\n",
    "query = \"\"\"\n",
    "        SELECT DISTINCT\n",
    "            dataFrame.*, Indian_ALLELE_FREQ_1, Indian_ALLELE_FREQ_2, Indian.ID\n",
    "        FROM\n",
    "            dataFrame LEFT OUTER JOIN \n",
    "            Indian ON dataFrame.Coordinate=Indian.POS \n",
    "            AND dataFrame.Chr=Indian.CHROM\n",
    "        \"\"\"\n",
    "df_I = pd.read_sql_query(query , conn)\n",
    "\n",
    "query = \"\"\"\n",
    "        SELECT DISTINCT\n",
    "            dataFrame.*, Malay_ALLELE_FREQ_1, Malay_ALLELE_FREQ_2, Malay.ID\n",
    "        FROM\n",
    "            dataFrame LEFT OUTER JOIN\n",
    "            Malay ON dataFrame.Coordinate=Malay.POS\n",
    "            AND dataFrame.Chr=Malay.CHROM\n",
    "        \"\"\"\n",
    "df_M = pd.read_sql_query(query , conn)\n",
    "\n",
    "query = \"\"\"\n",
    "        SELECT DISTINCT\n",
    "            dataFrame.*, Chinese_ALLELE_FREQ_1, Chinese_ALLELE_FREQ_2, Chinese_ALLELE_FREQ_3, Chinese_ALLELE_FREQ_4, Chinese.ID\n",
    "        FROM\n",
    "            dataFrame LEFT OUTER JOIN\n",
    "            Chinese ON dataFrame.Coordinate=Chinese.POS\n",
    "            AND dataFrame.Chr=Chinese.CHROM\n",
    "        \"\"\"\n",
    "\n",
    "df_C = pd.read_sql_query(query , conn)\n",
    "        \n",
    "timeTaken = time.time()-startTime \n",
    "print(datetime.datetime.now().strftime('%Y/%m/%d %H:%M:%S'), \n",
    "      '- Data Retrieval for all races completed: Took {} seconds to complete.'.format(timeTaken))\n",
    "\n",
    "# 2017/09/18 15:54:03 - Data Retrieval for all races completed: Took 189.973806142807 seconds to complete (only exome).\n",
    "# 2017/09/20 13:35:04 - Data Retrieval for all races completed: Took 219.73676705360413 seconds to complete (only exome).\n",
    "# 2017/09/26 14:44:25 - Data Retrieval for all races completed: Took 465.7713861465454 seconds to complete (whole genome)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2017/11/30 16:57:25 - Data Retrieval for all races completed: Took 0.03325986862182617 seconds to complete.\n"
     ]
    }
   ],
   "source": [
    "startTime = time.time()\n",
    "\n",
    "query = \"\"\"\n",
    "        SELECT DISTINCT\n",
    "            dataFrame.*, CMI.*\n",
    "        FROM\n",
    "            dataFrame LEFT OUTER JOIN \n",
    "            CMI ON dataFrame.Coordinate=CMI.POS \n",
    "            AND dataFrame.Chr=CMI.CHROM\n",
    "        \"\"\"\n",
    "df = pd.read_sql_query(query , conn)\n",
    "\n",
    "timeTaken = time.time()-startTime \n",
    "print(datetime.datetime.now().strftime('%Y/%m/%d %H:%M:%S'), \n",
    "      '- Data Retrieval for all races completed: Took {} seconds to complete.'.format(timeTaken))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Multithreading/cpus "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 4 cpu's available in this machine\n",
      "2017/11/21 11:35:34 - Data Retrieval for all races completed: Took 673.142520904541 seconds to complete.\n"
     ]
    }
   ],
   "source": [
    "startTime = time.time()\n",
    "\n",
    "print('There are', multiprocessing.cpu_count(), 'cpu\\'s available in this machine')\n",
    "\n",
    "responses={}\n",
    "responses_lock=threading.Lock()\n",
    "\n",
    "maxconnections = 8\n",
    "pool_sema = BoundedSemaphore(value=maxconnections)\n",
    "\n",
    "def task(fname):\n",
    "    pool_sema.acquire()\n",
    "    conn = sqlite3.connect('dbase_Sqlite')\n",
    "    if fname != 'Chinese':\n",
    "        query = \"\"\"\n",
    "            SELECT DISTINCT\n",
    "                dataFrame.*, \"\"\" + fname + \"\"\"_ALLELE_FREQ_1, \"\"\" + fname + \"\"\"_ALLELE_FREQ_2, \"\"\" + fname + \"\"\".ID\n",
    "            FROM\n",
    "                dataFrame LEFT OUTER JOIN \n",
    "                \"\"\" + fname + \"\"\" ON dataFrame.Coordinate=\"\"\" + fname + \"\"\".POS \n",
    "            AND dataFrame.Chr=\"\"\" + fname + \"\"\".CHROM\n",
    "            \"\"\"\n",
    "    else:\n",
    "        query = \"\"\"\n",
    "            SELECT DISTINCT\n",
    "                dataFrame.*, \"\"\" + fname + \"\"\"_ALLELE_FREQ_1, \"\"\" + fname + \"\"\"_ALLELE_FREQ_2, \"\"\" + fname + \"\"\"_ALLELE_FREQ_3, \"\"\" + fname + \"\"\"_ALLELE_FREQ_4, \"\"\" + fname + \"\"\".ID\n",
    "            FROM\n",
    "                dataFrame LEFT OUTER JOIN \n",
    "                \"\"\" + fname + \"\"\" ON dataFrame.Coordinate=\"\"\" + fname + \"\"\".POS \n",
    "            AND dataFrame.Chr=\"\"\" + fname + \"\"\".CHROM\n",
    "            \"\"\"\n",
    "    df = pd.read_sql_query(query , conn)\n",
    "    conn.close()\n",
    "    pool_sema.release()\n",
    "    responses_lock.acquire()\n",
    "    responses[fname] = df\n",
    "    responses_lock.release()\n",
    "\n",
    "pool = []\n",
    "\n",
    "#find sql files and spawn theads\n",
    "for fname in ['Chinese', 'Indian', 'Malay']:\n",
    "    #create new thread with task\n",
    "    thread = threading.Thread(target=task,args=(fname,))\n",
    "    thread.daemon = True\n",
    "    # store thread in pool \n",
    "    pool.append(thread)\n",
    "    #thread started\n",
    "    thread.start()\n",
    "\n",
    "#wait for all threads tasks done\n",
    "for thread in pool:\n",
    "    thread.join()\n",
    "\n",
    "timeTaken = time.time()-startTime \n",
    "print(datetime.datetime.now().strftime('%Y/%m/%d %H:%M:%S'), \n",
    "      '- Data Retrieval for all races completed: Took {} seconds to complete.'.format(timeTaken))\n",
    "\n",
    "# 2017/09/20 13:40:13 - Data Retrieval for all races completed: Took 309.63539934158325 seconds to complete (Exome only).\n",
    "# 2017/09/20 13:53:34 - Data Retrieval for all races completed: Took 274.6211278438568 seconds to complete (Exome only).\n",
    "### Slower than the single thread operation, probably due to the memory limitation, tested below 80 seconds with 32gb server instance.\n",
    "# 2017/09/26 14:56:43 - Data Retrieval for all races completed: Took 610.4648928642273 seconds to complete (Whole genome)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Merging the output tables into one table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'responses' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-0aac8f938647>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf_M\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresponses\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Malay'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mdf_I\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresponses\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Indian'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mdf_C\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresponses\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Chinese'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m df = df_I.merge(df_C[['index','Chinese_ALLELE_FREQ_1','Chinese_ALLELE_FREQ_2','Chinese_ALLELE_FREQ_3', 'Chinese_ALLELE_FREQ_4']], \n\u001b[1;32m      5\u001b[0m                  on=['index'], how='outer').merge(df_M[['index','Malay_ALLELE_FREQ_1', 'Malay_ALLELE_FREQ_2']], on=['index'], how='outer')\n",
      "\u001b[0;31mNameError\u001b[0m: name 'responses' is not defined"
     ]
    }
   ],
   "source": [
    "df_M = responses['Malay']\n",
    "df_I = responses['Indian']\n",
    "df_C = responses['Chinese']\n",
    "df = df_I.merge(df_C[['index','Chinese_ALLELE_FREQ_1','Chinese_ALLELE_FREQ_2','Chinese_ALLELE_FREQ_3', 'Chinese_ALLELE_FREQ_4']], \n",
    "                 on=['index'], how='outer').merge(df_M[['index','Malay_ALLELE_FREQ_1', 'Malay_ALLELE_FREQ_2']], on=['index'], how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variant</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Coordinate</th>\n",
       "      <th>Variant Length</th>\n",
       "      <th>Type</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Exonic</th>\n",
       "      <th>...</th>\n",
       "      <th>Chinese_ALLELE_FREQ_2</th>\n",
       "      <th>Chinese_ALLELE_FREQ_3</th>\n",
       "      <th>Chinese_ALLELE_FREQ_4</th>\n",
       "      <th>ID</th>\n",
       "      <th>N_CHR_Malay</th>\n",
       "      <th>Malay_ALLELE_FREQ_1</th>\n",
       "      <th>Malay_ALLELE_FREQ_2</th>\n",
       "      <th>N_CHR_Indian</th>\n",
       "      <th>Indian_ALLELE_FREQ_1</th>\n",
       "      <th>Indian_ALLELE_FREQ_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>T1-MS</td>\n",
       "      <td>None</td>\n",
       "      <td>A&gt;A/G</td>\n",
       "      <td>1</td>\n",
       "      <td>14804874</td>\n",
       "      <td>1</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>G:0.456989</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rs182380651</td>\n",
       "      <td>192.0</td>\n",
       "      <td>A:0.666667</td>\n",
       "      <td>G:0.333333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>T1-MS</td>\n",
       "      <td>None</td>\n",
       "      <td>A&gt;A/G</td>\n",
       "      <td>1</td>\n",
       "      <td>14804874</td>\n",
       "      <td>1</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>G:0.456989</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rs189939583</td>\n",
       "      <td>192.0</td>\n",
       "      <td>A:0.666667</td>\n",
       "      <td>G:0.333333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>T1-MS</td>\n",
       "      <td>None</td>\n",
       "      <td>A&gt;A/G</td>\n",
       "      <td>1</td>\n",
       "      <td>14804874</td>\n",
       "      <td>1</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>G:0.456989</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rs7555566</td>\n",
       "      <td>192.0</td>\n",
       "      <td>A:0.666667</td>\n",
       "      <td>G:0.333333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>T1-MS</td>\n",
       "      <td>SDHB</td>\n",
       "      <td>G&gt;T/T</td>\n",
       "      <td>1</td>\n",
       "      <td>17380497</td>\n",
       "      <td>1</td>\n",
       "      <td>snv</td>\n",
       "      <td>hom</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>T:1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rs2746462</td>\n",
       "      <td>192.0</td>\n",
       "      <td>G:0</td>\n",
       "      <td>T:1</td>\n",
       "      <td>72.0</td>\n",
       "      <td>G:0.0277778</td>\n",
       "      <td>T:0.972222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>T1-MS</td>\n",
       "      <td>MUTYH</td>\n",
       "      <td>C&gt;C/G</td>\n",
       "      <td>1</td>\n",
       "      <td>45797505</td>\n",
       "      <td>1</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>G:0.33871</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rs3219489</td>\n",
       "      <td>192.0</td>\n",
       "      <td>C:0.583333</td>\n",
       "      <td>G:0.416667</td>\n",
       "      <td>72.0</td>\n",
       "      <td>C:0.819444</td>\n",
       "      <td>G:0.180556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3</td>\n",
       "      <td>T1-MS</td>\n",
       "      <td>MUTYH</td>\n",
       "      <td>C&gt;C/G</td>\n",
       "      <td>1</td>\n",
       "      <td>45797505</td>\n",
       "      <td>1</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>G:0.33871</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rs3219489</td>\n",
       "      <td>192.0</td>\n",
       "      <td>C:0.583333</td>\n",
       "      <td>G:0.416667</td>\n",
       "      <td>72.0</td>\n",
       "      <td>C:0.819444</td>\n",
       "      <td>G:0.180556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4</td>\n",
       "      <td>T1-MS</td>\n",
       "      <td>None</td>\n",
       "      <td>C&gt;C/T</td>\n",
       "      <td>1</td>\n",
       "      <td>150860471</td>\n",
       "      <td>1</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>T:0.376344</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rs538465706</td>\n",
       "      <td>192.0</td>\n",
       "      <td>C:0.557292</td>\n",
       "      <td>T:0.442708</td>\n",
       "      <td>72.0</td>\n",
       "      <td>C:0.638889</td>\n",
       "      <td>T:0.361111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4</td>\n",
       "      <td>T1-MS</td>\n",
       "      <td>None</td>\n",
       "      <td>C&gt;C/T</td>\n",
       "      <td>1</td>\n",
       "      <td>150860471</td>\n",
       "      <td>1</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>T:0.376344</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rs7412746</td>\n",
       "      <td>192.0</td>\n",
       "      <td>C:0.557292</td>\n",
       "      <td>T:0.442708</td>\n",
       "      <td>72.0</td>\n",
       "      <td>C:0.638889</td>\n",
       "      <td>T:0.361111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5</td>\n",
       "      <td>T1-MS</td>\n",
       "      <td>NR5A2</td>\n",
       "      <td>A&gt;A/G</td>\n",
       "      <td>1</td>\n",
       "      <td>200007432</td>\n",
       "      <td>1</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>G:0.602151</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rs3790844</td>\n",
       "      <td>192.0</td>\n",
       "      <td>A:0.473958</td>\n",
       "      <td>G:0.526042</td>\n",
       "      <td>72.0</td>\n",
       "      <td>A:0.541667</td>\n",
       "      <td>G:0.458333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>T1-MS</td>\n",
       "      <td>None</td>\n",
       "      <td>A&gt;A/G</td>\n",
       "      <td>1</td>\n",
       "      <td>222164948</td>\n",
       "      <td>1</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>G:0.166667</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rs6687758</td>\n",
       "      <td>192.0</td>\n",
       "      <td>A:0.885417</td>\n",
       "      <td>G:0.114583</td>\n",
       "      <td>72.0</td>\n",
       "      <td>A:0.833333</td>\n",
       "      <td>G:0.166667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 95 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   index Sample   Gene Variant Chr  Coordinate  Variant Length Type Genotype  \\\n",
       "0      0  T1-MS   None   A>A/G   1    14804874               1  snv      het   \n",
       "1      0  T1-MS   None   A>A/G   1    14804874               1  snv      het   \n",
       "2      0  T1-MS   None   A>A/G   1    14804874               1  snv      het   \n",
       "3      1  T1-MS   SDHB   G>T/T   1    17380497               1  snv      hom   \n",
       "4      2  T1-MS  MUTYH   C>C/G   1    45797505               1  snv      het   \n",
       "5      3  T1-MS  MUTYH   C>C/G   1    45797505               1  snv      het   \n",
       "6      4  T1-MS   None   C>C/T   1   150860471               1  snv      het   \n",
       "7      4  T1-MS   None   C>C/T   1   150860471               1  snv      het   \n",
       "8      5  T1-MS  NR5A2   A>A/G   1   200007432               1  snv      het   \n",
       "9      6  T1-MS   None   A>A/G   1   222164948               1  snv      het   \n",
       "\n",
       "  Exonic         ...          Chinese_ALLELE_FREQ_2  Chinese_ALLELE_FREQ_3  \\\n",
       "0     no         ...                     G:0.456989                   None   \n",
       "1     no         ...                     G:0.456989                   None   \n",
       "2     no         ...                     G:0.456989                   None   \n",
       "3    yes         ...                            T:1                   None   \n",
       "4    yes         ...                      G:0.33871                   None   \n",
       "5    yes         ...                      G:0.33871                   None   \n",
       "6     no         ...                     T:0.376344                   None   \n",
       "7     no         ...                     T:0.376344                   None   \n",
       "8     no         ...                     G:0.602151                   None   \n",
       "9     no         ...                     G:0.166667                   None   \n",
       "\n",
       "   Chinese_ALLELE_FREQ_4           ID N_CHR_Malay  Malay_ALLELE_FREQ_1  \\\n",
       "0                   None  rs182380651       192.0           A:0.666667   \n",
       "1                   None  rs189939583       192.0           A:0.666667   \n",
       "2                   None    rs7555566       192.0           A:0.666667   \n",
       "3                   None    rs2746462       192.0                  G:0   \n",
       "4                   None    rs3219489       192.0           C:0.583333   \n",
       "5                   None    rs3219489       192.0           C:0.583333   \n",
       "6                   None  rs538465706       192.0           C:0.557292   \n",
       "7                   None    rs7412746       192.0           C:0.557292   \n",
       "8                   None    rs3790844       192.0           A:0.473958   \n",
       "9                   None    rs6687758       192.0           A:0.885417   \n",
       "\n",
       "   Malay_ALLELE_FREQ_2  N_CHR_Indian Indian_ALLELE_FREQ_1 Indian_ALLELE_FREQ_2  \n",
       "0           G:0.333333           NaN                 None                 None  \n",
       "1           G:0.333333           NaN                 None                 None  \n",
       "2           G:0.333333           NaN                 None                 None  \n",
       "3                  T:1          72.0          G:0.0277778           T:0.972222  \n",
       "4           G:0.416667          72.0           C:0.819444           G:0.180556  \n",
       "5           G:0.416667          72.0           C:0.819444           G:0.180556  \n",
       "6           T:0.442708          72.0           C:0.638889           T:0.361111  \n",
       "7           T:0.442708          72.0           C:0.638889           T:0.361111  \n",
       "8           G:0.526042          72.0           A:0.541667           G:0.458333  \n",
       "9           G:0.114583          72.0           A:0.833333           G:0.166667  \n",
       "\n",
       "[10 rows x 95 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_working = df[:]\n",
    "\n",
    "for ethnic in ['Indian', 'Malay', 'Chinese']:\n",
    "    df_Null_temp = df_working[df_working[ethnic + \"_ALLELE_FREQ_1\"].isnull()]\n",
    "    \n",
    "    for chrom in df_Null_temp.CHROM.unique():\n",
    "        df_Null_temp_working = df_Null_temp[(df_Null_temp.Chr.astype(str)==str(chrom))]\n",
    "        with open('query.bed','w') as bedFile:\n",
    "            for i in df_Null_temp_working['Coordinate']:\n",
    "                bedFile.write('chr' + str(chrom) + '\\t' + str(i-1) + '\\t' + str(i) + '\\n')\n",
    "        with open('output.fa', 'w') as outfile:\n",
    "            proc = subprocess.Popen(['seqtk', 'subseq', 'data/GRCh37/chr' + str(chrom) + '.fa', 'query.bed'], \n",
    "                                     stdin=subprocess.PIPE, stdout=outfile)\n",
    "            out, err = proc.communicate()\n",
    "        \n",
    "        p = open('output.fa', \"r\")\n",
    "        for i in df_Null_temp_working.index: \n",
    "            p.readline()\n",
    "            df_working.loc[(df_working.index==i),(ethnic + '_ALLELE_FREQ_1')] = str.capitalize(p.readline().replace('\\n', ':1'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variant</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Coordinate</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Type</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Exonic</th>\n",
       "      <th>...</th>\n",
       "      <th>Chinese_ALLELE_FREQ_2</th>\n",
       "      <th>Chinese_ALLELE_FREQ_3</th>\n",
       "      <th>Chinese_ALLELE_FREQ_4</th>\n",
       "      <th>ID</th>\n",
       "      <th>N_CHR_Malay</th>\n",
       "      <th>Malay_ALLELE_FREQ_1</th>\n",
       "      <th>Malay_ALLELE_FREQ_2</th>\n",
       "      <th>N_CHR_Indian</th>\n",
       "      <th>Indian_ALLELE_FREQ_1</th>\n",
       "      <th>Indian_ALLELE_FREQ_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>B4-61-S5</td>\n",
       "      <td>SDHC</td>\n",
       "      <td>C&gt;C/T</td>\n",
       "      <td>1</td>\n",
       "      <td>161326475</td>\n",
       "      <td>None</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C:1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C:1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>B4-61-S5</td>\n",
       "      <td>SDHC</td>\n",
       "      <td>C&gt;C/A</td>\n",
       "      <td>1</td>\n",
       "      <td>161326488</td>\n",
       "      <td>None</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C:1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C:1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>B5-66-S10</td>\n",
       "      <td>FANCA</td>\n",
       "      <td>C&gt;C/A</td>\n",
       "      <td>16</td>\n",
       "      <td>89839690</td>\n",
       "      <td>None</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>T:0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rs182288490</td>\n",
       "      <td>None</td>\n",
       "      <td>C:1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C:1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>B5-81-S11</td>\n",
       "      <td>ERCC2</td>\n",
       "      <td>C&gt;C/G</td>\n",
       "      <td>19</td>\n",
       "      <td>45855870</td>\n",
       "      <td>None</td>\n",
       "      <td>snv</td>\n",
       "      <td>het</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>A:0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rs149818919</td>\n",
       "      <td>None</td>\n",
       "      <td>C:1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C:1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 95 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   index     Sample   Gene Variant  Chr  Coordinate Classification Type  \\\n",
       "0      0   B4-61-S5   SDHC   C>C/T    1   161326475           None  snv   \n",
       "1      1   B4-61-S5   SDHC   C>C/A    1   161326488           None  snv   \n",
       "2      2  B5-66-S10  FANCA   C>C/A   16    89839690           None  snv   \n",
       "3      3  B5-81-S11  ERCC2   C>C/G   19    45855870           None  snv   \n",
       "\n",
       "  Genotype Exonic         ...          Chinese_ALLELE_FREQ_2  \\\n",
       "0      het    yes         ...                           None   \n",
       "1      het    yes         ...                           None   \n",
       "2      het    yes         ...                            T:0   \n",
       "3      het    yes         ...                            A:0   \n",
       "\n",
       "   Chinese_ALLELE_FREQ_3  Chinese_ALLELE_FREQ_4           ID  N_CHR_Malay  \\\n",
       "0                   None                   None         None         None   \n",
       "1                   None                   None         None         None   \n",
       "2                   None                   None  rs182288490         None   \n",
       "3                   None                   None  rs149818919         None   \n",
       "\n",
       "   Malay_ALLELE_FREQ_1  Malay_ALLELE_FREQ_2 N_CHR_Indian Indian_ALLELE_FREQ_1  \\\n",
       "0                  C:1                 None         None                  C:1   \n",
       "1                  C:1                 None         None                  C:1   \n",
       "2                  C:1                 None         None                  C:1   \n",
       "3                  C:1                 None         None                  C:1   \n",
       "\n",
       "  Indian_ALLELE_FREQ_2  \n",
       "0                 None  \n",
       "1                 None  \n",
       "2                 None  \n",
       "3                 None  \n",
       "\n",
       "[4 rows x 95 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_working[:9]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Saving the output tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_working.to_excel(\"output.xls\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Query individual snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>N_CHR</th>\n",
       "      <th>Malay_ALLELE_FREQ_1</th>\n",
       "      <th>Malay_ALLELE_FREQ_2</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>47601106</td>\n",
       "      <td>2</td>\n",
       "      <td>192</td>\n",
       "      <td>T:0.192708</td>\n",
       "      <td>C:0.807292</td>\n",
       "      <td>rs1126497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>97816327</td>\n",
       "      <td>2</td>\n",
       "      <td>174</td>\n",
       "      <td>T:0.965517</td>\n",
       "      <td>C:0.0344828</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>97816327</td>\n",
       "      <td>2</td>\n",
       "      <td>174</td>\n",
       "      <td>T:0.965517</td>\n",
       "      <td>C:0.0344828</td>\n",
       "      <td>rs6465657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>97816327</td>\n",
       "      <td>2</td>\n",
       "      <td>174</td>\n",
       "      <td>T:0.965517</td>\n",
       "      <td>C:0.0344828</td>\n",
       "      <td>rs77759206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>176637576</td>\n",
       "      <td>2</td>\n",
       "      <td>192</td>\n",
       "      <td>T:0.536458</td>\n",
       "      <td>C:0.463542</td>\n",
       "      <td>rs28932178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7</td>\n",
       "      <td>97816327</td>\n",
       "      <td>2</td>\n",
       "      <td>192</td>\n",
       "      <td>C:0.739583</td>\n",
       "      <td>T:0.260417</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>97816327</td>\n",
       "      <td>2</td>\n",
       "      <td>192</td>\n",
       "      <td>C:0.739583</td>\n",
       "      <td>T:0.260417</td>\n",
       "      <td>rs6465657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>97816327</td>\n",
       "      <td>2</td>\n",
       "      <td>192</td>\n",
       "      <td>C:0.739583</td>\n",
       "      <td>T:0.260417</td>\n",
       "      <td>rs77759206</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS  N_ALLELES  N_CHR Malay_ALLELE_FREQ_1 Malay_ALLELE_FREQ_2  \\\n",
       "0      2   47601106          2    192          T:0.192708          C:0.807292   \n",
       "1      2   97816327          2    174          T:0.965517         C:0.0344828   \n",
       "2      2   97816327          2    174          T:0.965517         C:0.0344828   \n",
       "3      2   97816327          2    174          T:0.965517         C:0.0344828   \n",
       "4      5  176637576          2    192          T:0.536458          C:0.463542   \n",
       "5      7   97816327          2    192          C:0.739583          T:0.260417   \n",
       "6      7   97816327          2    192          C:0.739583          T:0.260417   \n",
       "7      7   97816327          2    192          C:0.739583          T:0.260417   \n",
       "\n",
       "           ID  \n",
       "0   rs1126497  \n",
       "1           .  \n",
       "2   rs6465657  \n",
       "3  rs77759206  \n",
       "4  rs28932178  \n",
       "5           .  \n",
       "6   rs6465657  \n",
       "7  rs77759206  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "        SELECT DISTINCT\n",
    "            Malay.*\n",
    "        FROM\n",
    "            Malay\n",
    "        WHERE POS in ('97816327', '14804874', '176637576', '150860471', '47601106')\n",
    "        \"\"\"\n",
    "\n",
    "conn = sqlite3.connect('dbase_Sqlite')\n",
    "\n",
    "df_testing = pd.read_sql_query(query , conn)\n",
    "df_testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xls_file = pd.ExcelFile('output.xls')\n",
    "table = xls_file.parse('Sheet1')\n",
    "\n",
    "xls_file = pd.ExcelFile('data/output.xls')\n",
    "table2 = xls_file.parse('Sheet1')\n",
    "\n",
    "table.equals(table2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
